src/cpu/pred/btb_entry.hh

```diff
@@ -0,0 +1,262 @@
+template class IndexingPolicyTemplate<BTBTagTypes>;
+
+class BTBSetAssociative : public BTBIndexingPolicy
+{
+  public:
+    PARAMS(BTBSetAssociative);
+    using KeyType = BTBTagTypes::KeyType;
+
+    BTBSetAssociative(const Params &p) : BTBIndexingPolicy(p)
+    {
+        setNumThreads(p.numThreads);
+    }
+
+  protected:
+    /**
+     * Extract the set index for the instruction PC based on tid.
+     */
+    uint32_t
+    extractSet(const KeyType &key) const
+    {
+        return ((key.address >> setShift)
+                ^ (key.tid << (tagShift - setShift - log2NumThreads)))
+            & setMask;
+    }
+
+  public:
+    /**
+     * Find all possible entries for insertion and replacement of an address.
+     */
+    std::vector<ReplaceableEntry*>
+    getPossibleEntries(const KeyType &key) const override
+    {
+        return sets[extractSet(key)];
+    }
+
+    /**
+     * Set number of threads sharing the BTB
+     */
+    void
+    setNumThreads(unsigned num_threads)
+    {
+        log2NumThreads = log2i(num_threads);
+    }
+
+    Addr regenerateAddr(const KeyType &key,
+                        const ReplaceableEntry* entry) const override
+    {
+        return (key.address << tagShift) | (entry->getSet() << setShift);
+    }
+
+  private:
+    unsigned log2NumThreads;
+};
+
+namespace branch_prediction
+{
+
+class BTBEntry : public ReplaceableEntry
+{
+  public:
+    using IndexingPolicy = gem5::BTBIndexingPolicy;
+    using KeyType = gem5::BTBTagTypes::KeyType;
+    using TagExtractor = std::function<Addr(Addr)>;
+
+    /** Default constructor */
+    BTBEntry(TagExtractor ext)
+      : inst(nullptr), tid(0), extractTag(ext), valid(false), tag(MaxAddr)
+    {}
+
+    /** Update the target and instruction in the BTB entry */
+    void
+    update(const PCStateBase &_target,
+           StaticInstPtr _inst)
+    {
+        set(target, _target);
+        inst = _inst;
+    }
+
+    /**
+     * Checks if the given tag information corresponds to this entry's.
+     */
+    bool
+    match(const KeyType &key)
+    {
+        return isValid() && (getTag() == extractTag(key.address))
+            && (tid == key.tid);
+    }
+
+    /**
+     * Insert the block by assigning it a tag and marking it valid. Touches
+     * block if it hadn't been touched previously.
+     */
+    void
+    insert(const KeyType &key)
+    {
+        setValid();
+        setTag(extractTag(key.address));
+        tid = key.tid;
+    }
+
+    /** Copy constructor */
+    BTBEntry(const BTBEntry &other)
+    {
+        valid      = other.valid;
+        tag        = other.tag;
+        tid        = other.tid;
+        inst       = other.inst;
+        extractTag = other.extractTag;
+        set(target, other.target);
+    }
+
+    /**
+     * Checks if the entry is valid.
+     */
+    virtual bool isValid() const { return valid; }
+
+    /**
+     * Get tag associated to this block.
+     */
+    virtual Addr getTag() const { return tag; }
```

Comment:
I think that it should instead return KeyType, since the tid *is* part of the tag. We need to make sure that what happened with the prefetchers (some use the secure bit, and others don´t) doesn´t happen in this class. We can already see an instance of the bug happening in this class´ print() and invalidate(), where the tid was forgotten.

Reply:
Changed the tag type to be KeyType througout.